File: Abot/librs/src/cfx.rs
use clap::Parser;
use playwright::Playwright;
use shared_memory::{ShmemConf, Shmem};
use tokio::time::{sleep, Duration};

#[derive(Parser)]
struct Args {
    #[arg(long)]
    shm_name: String,
    #[arg(long, default_value_t = false)]
    headless: bool,
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let shmem = ShmemConf::new()
        .os_id(&args.shm_name)
        .size(8)
        .create()?;
    let playwright = Playwright::initialize().await?;
    let chromium = playwright.chromium()?;
    let browser = chromium
        .launcher()
        .headless(args.headless)
        .launch()
        .await?;
    let context = browser.new_context_builder().build().await?;
    let page = context.new_page().await?;
    page.goto_builder("https://cfbenchmarks.com/data/assets/BTC")
        .goto()
        .await?;
    let pointer = r#"/html/body/div/div/main/h1/div[4]/div/span[1]/span"#;
    let locator = page
        .locator_builder(format!("xpath={}", pointer))
        .build();
    locator.wait_for_builder().wait_for().await?;
    loop {
        if let Some(raw) = locator.text_content().await? {
            if raw != "-" {
                let cleaned = raw.replace('$', "").replace(',', "");
                if let Ok(price) = cleaned.parse::<f64>() {
                    let bytes = price.to_le_bytes();
                    unsafe {
                        let buf = shmem.as_slice_mut();
                        buf.copy_from_slice(&bytes);
                    }
                }
            }
        }
        sleep(Duration::from_secs(1)).await;
    }
}

File: Abot/lib/cfx.py
from utils import get_set, get_db, ts, init_db
from playwright.async_api import async_playwright as ap
import asyncio

POINTER = '/html/body/div/div/main/h1/div[4]/div/span[1]/span'
LAST = None

async def insert(conn, value:float):
    """Insert price data into the database"""
    await conn.execute("""INSERT OR REPLACE INTO prices (time, value) VALUES (?, ?)""", (ts(), value))
    await conn.commit()

async def fetcher() -> str:
    """Spinup playwright and extract price data"""
    init_db()
    IDX, STATE, CFG= get_set()
    conn = await get_db()
    async with ap() as p:
        browser = await p.chromium.launch(headless=False)
        page = await browser.new_page()
        await page.goto("https://cfbenchmarks.com/data/assets/BTC")
        await page.wait_for_selector("xpath=" + POINTER)
        try:
            while True:
                txt = await page.locator("xpath=" + POINTER).text_content()
                if txt != '-' and txt is not None and txt != 0:
                    price = float(txt.replace("$", "").replace(",", ""))
                    if price != LAST:
                        print(price)
                        STATE[IDX.BTC] = price
                        LAST = price
                    await insert(conn, price)
                else:
                    pass
                await asyncio.sleep(CFG.CFX_FETCH / 1000)
        except KeyboardInterrupt:
            pass
        finally:
            await conn.close()
            await browser.close()

asyncio.run(fetcher())
